sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathClient.java

```diff
@@ -808,13 +1046,162 @@ public AccessControlChangeResult setAccessControlRecursive(List<PathAccessContro
+
+            changes.setBatchFailures(
+                response.getValue().getFailedEntries()
+                    .stream()
+                    .map(aclFailedEntry -> new AccessControlChangeFailure()
+                        .setDirectory(aclFailedEntry.getType().equals("DIRECTORY"))
+                        .setName(aclFailedEntry.getName())
+                        .setErrorMessage(aclFailedEntry.getErrorMessage())
+                    ).collect(Collectors.toList())
+            );
+
+            changes.setBatchCounters(new AccessControlChangeCounters()
+                .setChangedDirectoriesCount(response.getValue().getDirectoriesSuccessful())
+                .setChangedFilesCount(response.getValue().getFilesSuccessful())
+                .setFailedChangesCount(response.getValue().getFailureCount()));
+
+            changes.setAggregateCounters(new AccessControlChangeCounters()
+                .setChangedDirectoriesCount(directoriesSuccessfulCount.get())
+                .setChangedFilesCount(filesSuccessfulCount.get())
+                .setFailedChangesCount(failureCount.get()));
+
+            progressHandler.accept(
+                new ResponseBase<>(response.getRequest(), response.getStatusCode(), response.getHeaders(), changes,
+                    response.getDeserializedHeaders()));
+        }
+
+        /*
+        Determine if we are finished either because there is no new continuation (failure or finished) token or we have
+        hit maxBatches.
+         */
+        if ((newToken == null || newToken.isEmpty()) || (maxBatches != null && batchesCount.get() >= maxBatches)) {
+            AccessControlChangeResult result = new AccessControlChangeResult()
+                .setBatchFailures(batchFailures)
+                .setContinuationToken(effectiveNextToken)
+                .setCounters(new AccessControlChangeCounters()
+                    .setChangedDirectoriesCount(directoriesSuccessfulCount.get())
+                    .setChangedFilesCount(filesSuccessfulCount.get())
+                    .setFailedChangesCount(failureCount.get()));
+
+            return new ResponseBase<>(response.getRequest(), response.getStatusCode(), response.getHeaders(),
+                result, response.getDeserializedHeaders());
+        }
+
+        // If we're not finished, issue another request
+        try {
+            Callable<ResponseBase<PathsSetAccessControlRecursiveHeaders, SetAccessControlRecursiveResponse>> operation =
+                () -> this.dataLakeStorage.getPaths().setAccessControlRecursiveWithResponse(mode, null, effectiveNextToken,
+                    continueOnFailure, batchSize, accessControlStr, null, context);
+
+            ResponseBase<PathsSetAccessControlRecursiveHeaders, SetAccessControlRecursiveResponse> response2 =
+                StorageImplUtils.sendRequest(operation, timeout, DataLakeStorageException.class);
+
+            return setAccessControlRecursiveWithResponseHelper(response2, maxBatches,
```

Comment:
The timeout logic will be different in the sync client than the async client, though I don't know how much we care. The API in the async client is `Mono<Response<AccessControlChangeResult>>`, so if the customer wanted to only wait so long before the operation timed out they would use `Mono.block(Duration)` which would apply to all service requests being made (ex, if this needed to make 10 service calls that timeout would be for all ten service calls.)

The way the timeout is being used here in the synchronous client is for each individual call, so it times out differently. Personally, I'm fine with this as I think this is the better design, but if we want to make sure sync and async behave the same we'll need to slightly change the timeout handling logic here.

Reply:
Implementation of setAccessControlList was not producing correct results. Will follow up with the correct implementation in the following PR.